// fp-ts/lib/Const.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var function_1 = require("./function");
exports.URI = 'Const';
var Const = /** @class */ (function () {
    function Const(value) {
        this.value = value;
    }
    Const.prototype.map = function (f) {
        return this;
    };
    Const.prototype.contramap = function (f) {
        return this;
    };
    Const.prototype.fold = function (f) {
        return f(this.value);
    };
    Const.prototype.equals = function (S, fy) {
        return this.fold(function (x) { return fy.fold(function (y) { return S.equals(x)(y); }); });
    };
    Const.prototype.inspect = function () {
        return this.toString();
    };
    Const.prototype.toString = function () {
        return "new Const(" + function_1.toString(this.value) + ")";
    };
    return Const;
}());
exports.Const = Const;
exports.equals = function (S) { return function (fx) { return function (fy) {
    return fx.equals(S, fy);
}; }; };
exports.getSetoid = function (S) { return ({
    equals: function (x) { return function (y) { return exports.equals(S)(x)(y); }; }
}); };
exports.map = function (f, fa) { return fa.map(f); };
exports.contramap = function (fa) { return function (f) {
    return fa.contramap(f);
}; };
exports.getApply = function (S) { return ({
    URI: exports.URI,
    map: exports.map,
    ap: function (fab, fa) {
        return new Const(S.concat(fab.fold(function_1.identity))(fa.fold(function_1.identity)));
    }
}); };
exports.getApplicative = function (M) {
    var empty = new Const(M.empty());
    return __assign({}, exports.getApply(M), { of: function (b) {
            return empty;
        } });
};
exports.const_ = {
    URI: exports.URI,
    map: exports.map,
    contramap: exports.contramap
};
