// fp-ts/lib/Filterable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Either_1 = require("./Either");
var Option_1 = require("./Option");
var function_1 = require("./function");
var Ops = /** @class */ (function () {
    function Ops() {
    }
    Ops.prototype.partition = function (F) {
        return function (predicate) { return function (fa) {
            var _a = F.partitionMap(Either_1.fromPredicate(predicate, function_1.identity), fa), left = _a.left, right = _a.right;
            return { no: left, yes: right };
        }; };
    };
    Ops.prototype.filterMap = function (F) {
        return function (f) { return function (fa) {
            return F.partitionMap(function (a) { return f(a).fold(function () { return Either_1.left(null); }, function (b) { return Either_1.right(b); }); }, fa).right;
        }; };
    };
    Ops.prototype.filter = function (F) {
        var _this = this;
        return function (predicate) { return function (fa) { return _this.filterMap(F)(Option_1.fromPredicate(predicate))(fa); }; };
    };
    Ops.prototype.partitioned = function (F) {
        return function (fa) { return F.partitionMap(function (a) { return a; }, fa); };
    };
    Ops.prototype.filtered = function (F) {
        var _this = this;
        return function (fa) { return _this.filterMap(F)(function (oa) { return oa; })(fa); };
    };
    return Ops;
}());
exports.Ops = Ops;
var ops = new Ops();
exports.partition = ops.partition;
exports.filterMap = ops.filterMap;
exports.filter = ops.filter;
exports.partitioned = ops.partitioned;
exports.filtered = ops.filtered;
