// fp-ts/lib/Functor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var function_1 = require("./function");
var Ops = /** @class */ (function () {
    function Ops() {
    }
    Ops.prototype.lift = function (F) {
        return function (f) { return function (fa) { return F.map(f, fa); }; };
    };
    Ops.prototype.voidRight = function (F) {
        return function (a) { return function (fb) { return F.map(function_1.constant(a), fb); }; };
    };
    Ops.prototype.voidLeft = function (F) {
        return function (fa) { return function (b) { return F.map(function_1.constant(b), fa); }; };
    };
    Ops.prototype.flap = function (functor) {
        return function (ff) { return function (a) { return functor.map(function (f) { return f(a); }, ff); }; };
    };
    Ops.prototype.getFunctorComposition = function (F, G) {
        return {
            map: function (f, fa) { return F.map(function (ga) { return G.map(f, ga); }, fa); }
        };
    };
    return Ops;
}());
exports.Ops = Ops;
var ops = new Ops();
exports.lift = ops.lift;
exports.voidRight = ops.voidRight;
exports.voidLeft = ops.voidLeft;
exports.flap = ops.flap;
exports.getFunctorComposition = ops.getFunctorComposition;
