// fp-ts/lib/IxIO.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var io = require("./IO");
exports.URI = 'IxIO';
var IxIO = /** @class */ (function () {
    function IxIO(value) {
        this.value = value;
    }
    IxIO.prototype.run = function () {
        return this.value.run();
    };
    IxIO.prototype.ichain = function (f) {
        return new IxIO(this.value.chain(function (a) { return f(a).value; }));
    };
    IxIO.prototype.map = function (f) {
        return new IxIO(this.value.map(f));
    };
    IxIO.prototype.ap = function (fab) {
        return new IxIO(this.value.ap(fab.value));
    };
    IxIO.prototype.chain = function (f) {
        return new IxIO(this.value.chain(function (a) { return f(a).value; }));
    };
    return IxIO;
}());
exports.IxIO = IxIO;
exports.iof = function (a) { return new IxIO(io.of(a)); };
exports.ichain = function (f, fa) { return fa.ichain(f); };
exports.map = function (f, fa) { return fa.map(f); };
exports.of = exports.iof;
exports.ap = function (fab, fa) { return fa.ap(fab); };
exports.chain = function (f, fa) { return fa.chain(f); };
exports.ixIO = {
    URI: exports.URI,
    map: exports.map,
    of: exports.of,
    ap: exports.ap,
    chain: exports.chain,
    iof: exports.iof,
    ichain: exports.ichain
};
