// fp-ts/lib/ReaderT.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Ops = /** @class */ (function () {
    function Ops() {
    }
    Ops.prototype.map = function (F) {
        return function (f, fa) { return function (e) { return F.map(f, fa(e)); }; };
    };
    Ops.prototype.of = function (F) {
        return function (a) { return function (e) { return F.of(a); }; };
    };
    Ops.prototype.ap = function (F) {
        return function (fab, fa) { return function (e) { return F.ap(fab(e), fa(e)); }; };
    };
    Ops.prototype.chain = function (F) {
        return function (f, fa) { return function (e) { return F.chain(function (a) { return f(a)(e); }, fa(e)); }; };
    };
    Ops.prototype.ask = function (F) {
        return function () { return function (e) { return F.of(e); }; };
    };
    Ops.prototype.asks = function (F) {
        return function (f) { return function (e) { return F.of(f(e)); }; };
    };
    Ops.prototype.getReaderT = function (M) {
        return {
            map: this.map(M),
            of: this.of(M),
            ap: this.ap(M),
            chain: this.chain(M)
        };
    };
    return Ops;
}());
exports.Ops = Ops;
var ops = new Ops();
exports.map = ops.map;
exports.of = ops.of;
exports.ap = ops.ap;
exports.chain = ops.chain;
exports.ask = ops.ask;
exports.asks = ops.asks;
exports.getReaderT = ops.getReaderT;
