// fp-ts/lib/Reader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var function_1 = require("./function");
exports.URI = 'Reader';
var Reader = /** @class */ (function () {
    function Reader(run) {
        this.run = run;
    }
    Reader.prototype.map = function (f) {
        var _this = this;
        return new Reader(function (e) { return f(_this.run(e)); });
    };
    Reader.prototype.ap = function (fab) {
        var _this = this;
        return new Reader(function (e) { return fab.run(e)(_this.run(e)); });
    };
    Reader.prototype.chain = function (f) {
        var _this = this;
        return new Reader(function (e) { return f(_this.run(e)).run(e); });
    };
    return Reader;
}());
exports.Reader = Reader;
exports.map = function (f, fa) { return fa.map(f); };
exports.of = function (a) { return new Reader(function (e) { return a; }); };
exports.ap = function (fab, fa) { return fa.ap(fab); };
exports.chain = function (f, fa) { return fa.chain(f); };
/** reads the current context */
exports.ask = function () { return new Reader(function_1.identity); };
/** Projects a value from the global context in a Reader */
exports.asks = function (f) { return new Reader(f); };
/** changes the value of the local context during the execution of the action `fa` */
exports.local = function (f) { return function (fa) {
    return new Reader(function (e) { return fa.run(f(e)); });
}; };
exports.reader = {
    URI: exports.URI,
    map: exports.map,
    of: exports.of,
    ap: exports.ap,
    chain: exports.chain
};
