// fp-ts/lib/TaskEither.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var eitherT = require("./EitherT");
var task = require("./Task");
var eitherTTask = eitherT.getEitherT(task);
exports.URI = 'TaskEither';
var TaskEither = /** @class */ (function () {
    function TaskEither(value) {
        this.value = value;
    }
    TaskEither.prototype.map = function (f) {
        return new TaskEither(eitherTTask.map(f, this.value));
    };
    TaskEither.prototype.ap = function (fab) {
        return new TaskEither(eitherTTask.ap(fab.value, this.value));
    };
    TaskEither.prototype.ap_ = function (fb) {
        return fb.ap(this);
    };
    TaskEither.prototype.chain = function (f) {
        return new TaskEither(eitherTTask.chain(function (a) { return f(a).value; }, this.value));
    };
    TaskEither.prototype.fold = function (left, right) {
        return eitherT.fold(task)(left, right, this.value);
    };
    TaskEither.prototype.mapLeft = function (f) {
        return new TaskEither(eitherT.mapLeft(task)(f)(this.value));
    };
    TaskEither.prototype.toOption = function () {
        return eitherT.toOption(task)(this.value);
    };
    return TaskEither;
}());
exports.TaskEither = TaskEither;
exports.map = function (f, fa) { return fa.map(f); };
exports.of = function (a) { return new TaskEither(eitherTTask.of(a)); };
exports.ap = function (fab, fa) { return fa.ap(fab); };
exports.chain = function (f, fa) { return fa.chain(f); };
exports.right = function (fa) { return new TaskEither(eitherT.right(task)(fa)); };
exports.left = function (fa) { return new TaskEither(eitherT.left(task)(fa)); };
exports.fromEither = function (fa) {
    return new TaskEither(eitherT.fromEither(task)(fa));
};
exports.tryCatch = function (f, onrejected) {
    return new TaskEither(task.tryCatch(f)(onrejected));
};
exports.taskEither = {
    URI: exports.URI,
    map: exports.map,
    of: exports.of,
    ap: exports.ap,
    chain: exports.chain
};
